managed/services/telemetry/telemetry_test.go

```diff
@@ -0,0 +1,363 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/DATA-DOG/go-sqlmock"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	reporter "github.com/percona-platform/saas/gen/telemetry/reporter"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"gopkg.in/reform.v1"
+	"gopkg.in/reform.v1/dialects/postgresql"
+
+	"github.com/percona/pmm/api/serverpb"
+	"github.com/percona/pmm/managed/models"
+)
+
+func TestRunTelemetryService(t *testing.T) {
+	type fields struct {
+		db                  func() *reform.DB
+		l                   *logrus.Entry
+		start               time.Time
+		config              ServiceConfig
+		dsRegistry          func() DataSourceLocator
+		pmmVersion          string
+		os                  string
+		sDistributionMethod serverpb.DistributionMethod
+		tDistributionMethod pmmv1.DistributionMethod
+		dus                 func(l *logrus.Entry) distributionUtilService
+	}
+	const testSourceName = "PMMDB_SELECT"
+	const pmmVersion = "2.29"
+
+	now := time.Now()
+
+	expectedServerMetrics := []*pmmv1.ServerMetric_Metric{
+		{
+			Key:   "key",
+			Value: "value",
+		},
+		{
+			Key:   "key2",
+			Value: "value2",
+		},
+		{
+			Key:   "key3",
+			Value: "value3",
+		},
+	}
+	expectedReport := &reporter.ReportRequest{
+		Metrics: []*pmmv1.ServerMetric{
+			{
+				DistributionMethod: pmmv1.DistributionMethod_AMI,
+				Metrics:            expectedServerMetrics,
+			},
+		},
+	}
+	tests := []struct {
+		name                string
+		fields              fields
+		mockTelemetrySender func() sender
+		testTimeout         time.Duration
+	}{
+		{
+			name:        "should send metrics only once during start",
+			testTimeout: 2 * time.Second,
+			fields: fields{
+				db:         initMockDB(t, now, 1),
+				start:      now,
+				config:     getTestConfig(true, testSourceName, 10*time.Second),
+				dsRegistry: mockDataSourceLocator(t, [][]*pmmv1.ServerMetric_Metric{expectedServerMetrics}, testSourceName, 1),
+				pmmVersion: pmmVersion,
+				dus: func(l *logrus.Entry) distributionUtilService {
+					var dusMock mockDistributionUtilService
+					dusMock.Test(t)
+					dusMock.On("getDistributionMethodAndOS", l).Return(serverpb.DistributionMethod_AMI, pmmv1.DistributionMethod_AMI, "ami").
+						Times(1)
```

Comment:
To be honest I think there is no need to mock this function at all. We can just make `distributionInfoFilePath` configurable and use in tests tmp files with prepared content . In that case code will be simpler and tests will actually test that we detect distribution method correctly.

Reply:
It's whant I wanted to do at first, but it's not possible to create this file in OS by user, it returns an error that `/srv/pmm-distribution` file can be read only

Reply:
You can generate `tmp` files, and overwrite `distributionInfoFilePath`.